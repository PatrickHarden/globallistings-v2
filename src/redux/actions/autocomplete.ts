import { createStandardAction } from 'typesafe-actions';
import { State } from '../../types/state';
import { autocompleteSelector } from '../selectors/state';
import { updatePath } from './navigation';

export const selectedAutocompletionItem = createStandardAction('SELECT_AUTOCOMPLETION_ITEM')<
  string
>();

export const updateUrlWithSelectedLocation = (placeId: string) => (
  dispatch: Function,
  getState: () => State
) => {
  const place = autocompleteSelector(getState())[placeId];
  if (!!place) {
    dispatch(updatePath('location', place.description));
  }
};
