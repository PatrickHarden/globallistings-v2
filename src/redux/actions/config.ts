import axios from 'axios';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators-singular';
import { Config } from '../../types/config';
import { State } from '../../types/state';
import { configEndpointForCurrentUsageTypeSelector } from '../selectors/plp';

export const createFetchConfigRequest = createRequestPending('FETCH_CONFIG_REQUEST');
export const createFetchConfigFailure = createRequestFailure('FETCH_CONFIG_FAILURE');
export const createFetchConfigSuccess = createRequestSuccess<Config, 'FETCH_CONFIG_SUCCESS'>(
  'FETCH_CONFIG_SUCCESS'
);

// This will be used in the future to switch between multiple configs
export const fetchConfig = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const configEndpoint = configEndpointForCurrentUsageTypeSelector(state);

  dispatch(createFetchConfigRequest());
  return axios
    .get(configEndpoint)
    .then(res => {
      dispatch(createFetchConfigSuccess(res.data));
    })
    .catch(ex => {
      dispatch(createFetchConfigFailure(ex));
    });
};
