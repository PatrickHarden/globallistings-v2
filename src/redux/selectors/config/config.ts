import { createSelector } from 'reselect';
import { configSelector, configMetaSelector } from '../state';

export const configStatusSelector = createSelector(
  configMetaSelector,
  config => config.status
);

export const getFilters = createSelector(
  configSelector,
  config => config.filters
);

export const siteTypeSelector = createSelector(
  configSelector,
  config => config.siteType
);
export const searchModeSelector = createSelector(
  configSelector,
  config => config.searchMode
);

// SearchConfig
export const searchConfigSelector = createSelector(
  configSelector,
  config => config.searchConfig
);

export const searchBiasSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchBias
);

export const searchRadiusRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.biasRadius
);

export const searchCountryRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.restrictToCountry
);

export const polygonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.polygons
);

export const searchTypesRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPlaceTypes
);

export const assumeZipCodeSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.allowAssumeZipCode
);

export const searchPathSortAlphabeticalSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSortAlphabetical
);

export const searchPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSelector
);

export const propertyIdPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.propertyIdSearchPathSelector
);

export const hideBuyButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToBuy
);

export const hideLetButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToLet
);

export const getSearchHeader = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchHeader
);
export const searchResultsPageSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchResultsPage
);

export const searchSuggestDebounceSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchSuggestDebounce
);

export const defaultSearchLocationNameSelector = createSelector(
  configSelector,
  config => config.searchLocationName
);

export const paramsSelector = createSelector(
  configSelector,
  config => config.params
);

export const restrictionsSelector = createSelector(
  configSelector,
  config => ({
    country: config.searchConfig.restrictToCountry,
  })
);

export const otherConfigsSelector = createSelector(
  configSelector,
  config => (config.otherConfigs ? config.otherConfigs : [])
);

export const currentConfigSelector = createSelector(
  configSelector,
  config => config.currentConfig
);
