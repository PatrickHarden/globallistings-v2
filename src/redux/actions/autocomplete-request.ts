import { placesAutocomplete } from '../../api/google-maps/places';
import { State } from '../../types/state';
import { placeAutoCompleteConfig } from '../selectors/config/autocomplete-config';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators';
import { normaliseAutocompleteResponse } from '../../api/normalisers';
import { NormalisedAutocompleteResult } from '../../types/autocomplete';

export const suggestionsRequest = createRequestPending('GET_SUGGESTIONS_REQUEST');
export const suggestionsFailure = createRequestFailure('GET_SUGGESTIONS_FAILURE');
export const suggestionsSuccess = createRequestSuccess<
  NormalisedAutocompleteResult,
  'GET_SUGGESTIONS_SUCCESS'
>('GET_SUGGESTIONS_SUCCESS');

export const lookupPlaceByName = (name: string) => (dispatch: Function, getState: () => State) => {
  dispatch(suggestionsRequest(name));
  const state = getState();
  const config = placeAutoCompleteConfig(state);

  return placesAutocomplete(name, config)
    .then(suggestions => {
      const normalisedData = normaliseAutocompleteResponse(suggestions);
      dispatch(suggestionsSuccess(normalisedData, name));
    })
    .catch(ex => {
      dispatch(suggestionsFailure(ex, name));
    });
};
