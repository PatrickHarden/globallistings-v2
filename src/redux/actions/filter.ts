import { FilterModifier } from '../../types/config';
import { updateQueryParams, updatePath } from './navigation';
import { isPlpPathValue } from '../../types/plp';

export const handleFilterChange = (param: string, value: string, modifier: FilterModifier) => (
  dispatch: Function
) => {
  if (modifier === FilterModifier.url) {
    if (!isPlpPathValue(param)) {
      throw new Error(
        `Can only modify path for predefined values. Value ${param} is not supported`
      );
    }
    dispatch(updatePath(param, value));
  } else {
    dispatch(updateQueryParams({ [param]: value }));
  }
};
