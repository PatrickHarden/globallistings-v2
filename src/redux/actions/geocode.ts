import { placeGeocode } from '../../api/google-maps/places';
import { State, Geocode } from '../../types/state';
import { restrictionsSelector } from '../selectors/config/config';

import { plpLocationNameSelector } from '../selectors/location';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators';

export const createGeocodeRequest = createRequestPending('GEOCODE_REQUEST');
export const createGeocodeFailure = createRequestFailure('GEOCODE_FAILURE');
export const createGeocodeSuccess = createRequestSuccess<Geocode, 'GEOCODE_SUCCESS'>(
  'GEOCODE_SUCCESS'
);

export const validateUrl = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const placeName = plpLocationNameSelector(state);
  const restrictions = restrictionsSelector(state);

  dispatch(createGeocodeRequest(placeName));

  return placeGeocode(placeName, restrictions)
    .then(place => {
      dispatch(createGeocodeSuccess(place, placeName));
    })
    .catch(err => dispatch(createGeocodeFailure(err, placeName)));
};
