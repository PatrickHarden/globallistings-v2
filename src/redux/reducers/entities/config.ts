import { convert } from '../../../api/converters/config';
import { createFetchConfigSuccess } from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { DeepPartial } from 'redux';
import { Config } from '../../../types/config';

export type Actions = ActionType<typeof createFetchConfigSuccess>;

export default (state: DeepPartial<Config> = {}, action: Actions): Config | {} => {
  switch (action.type) {
    case 'FETCH_CONFIG_SUCCESS':
      return convert(action.payload);
    default:
      return state;
  }
};
