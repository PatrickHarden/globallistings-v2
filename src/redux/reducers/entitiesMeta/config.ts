import * as actions from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { handleError, handleSuccess, handlePending } from './shared/meta-reducers';

export type Actions = ActionType<typeof actions>;

export default (state = {}, action: Actions) => {
  switch (action.type) {
    case 'FETCH_CONFIG_REQUEST':
      return handlePending();
    case 'FETCH_CONFIG_SUCCESS':
      return handleSuccess(action);
    case 'FETCH_CONFIG_FAILURE':
      return handleError(action);
    default:
      return state;
  }
};
