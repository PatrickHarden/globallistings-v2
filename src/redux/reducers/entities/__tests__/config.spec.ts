import config from '../config';
import { createFetchConfigSuccess } from '../../../actions/config';
import { ActionType } from 'typesafe-actions';

describe(config, () => {
  it('unrelated action return same state', () => {
    const state = { foo: 'zap' };
    const action = {
      type: 'UNRELATED',
      payload: {},
    };
    const newState = config(state, action as ActionType<typeof createFetchConfigSuccess>);
    expect(newState).toEqual(state);
  });

  it('should return default state if undefined state is passed', () => {
    const action = {
      type: 'UNRELATED',
      payload: {},
    };
    const newState = config(undefined, action as ActionType<typeof createFetchConfigSuccess>);
    expect(newState).toEqual({});
  });

  it('should handle FetchConfigSuccessAction action ', () => {
    const action = {
      type: 'FETCH_CONFIG_SUCCESS',
      payload: {
        i18n: {
          one: 'uno',
        },
        searchConfig: {
          searchResultsPage: 'zap',
        },
      },
    } as ActionType<typeof createFetchConfigSuccess>;
    const newState = config(undefined, action);
    expect(newState).toMatchSnapshot();
  });
});
