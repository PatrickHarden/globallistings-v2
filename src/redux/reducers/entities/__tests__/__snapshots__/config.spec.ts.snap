// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_default should handle FetchConfigSuccessAction action  1`] = `
Object {
  "i18n": Object {
    "one": "uno",
  },
  "searchConfig": Object {
    "searchResultsPage": "zap",
  },
}
`;
